chore: Theia 1.37.0

kittaakos opened the pull request:
### Motivation
<!-- Why this pull request? -->

 - To use the latest version of the Theia framework,
 - To get rid of all `yarn audit` vulnerabilities,
 - To use the latest `electron` with `contextIsolation` enabled and `nodeIntegration` disabled, and
 - To enable the new high-contrast light theme.

### Change description
<!-- What does your code do? -->

 - Updated `@theia/*` to `1.37.0`.
 - Fixed all `yarn audit` security vulnerabilities.
 - Updated to `electron@23.2.4`:
   - `contextIsolation` is `true`,
   - `nodeIntegration` is `false`, and the
   - `webpack` target is moved from `electron-renderer` to `web`.
 - Updated to `typescript@4.9.3`.
 - Updated the `eslint` plugins.
 - Added a new light high contrast theme to the IDE2.
 - High contrast themes use Theia APIs for style adjustments.
 - Support for ESM modules: `"moduleResolution": "node16"`.
 - Node.js >= 16.14 is required.
 - VISX language packs were bumped to `1.70.0`.
 - Removed undesired editor context menu items.
 - [build: use `execFileSync` for npm scripts](https://github.com/arduino/arduino-ide/pull/2027/commits/c4bcdede5ade6c744cb659597aa6b448692e7934)
 - [build: use `execa` for the packager](https://github.com/arduino/arduino-ide/pull/2027/commits/c155a33eb8f8933bf7473f6a557a19fd99a68a3d)

----

```sh
% yarn audit
yarn audit v1.22.18
0 vulnerabilities found - Packages audited: 2091
✨  Done in 1.44s.
```

`Light High Contrast` theme:

https://user-images.githubusercontent.com/1405703/234628350-08cf8f3b-9585-4532-ab74-4bbebdbf11c9.mp4



### Other information
<!-- Any additional information that could help the review process -->

The application is not signed. On macOS, follow [these](https://github.com/arduino/arduino-ide/blob/main/docs/contributor-guide/beta-testing.md#the-risky-approach) steps.

Closes arduino/arduino-ide#321
Closes arduino/arduino-ide#1394

![image](https://user-images.githubusercontent.com/1405703/234807613-f2b2b2a0-85f9-46ac-9c74-965f6e7d03d5.png)
![image](https://user-images.githubusercontent.com/1405703/235104056-b6d985a1-3e78-49a4-9b62-2f92d126b5e6.png)



### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

kittaakos commented:
Upstream: https://github.com/sindresorhus/cpy/issues/109

per1234 reviewed (changes requested):
## Unexpected change to `workbench.iconTheme` default

The default value of the `workbench.iconTheme` setting changed from `none` to `theia-file-icons`.

[I don't have a strong opinion on the subject](https://github.com/arduino/arduino-ide/issues/1173#issuecomment-1192579246) of what the default should be. However, the decision the last time the subject was discussed ((https://github.com/arduino/arduino-ide/issues/1173) was that the default should be `none` and I am not aware of any change to the decision since then so I assume this was an inadvertent change.

### To reproduce

1. Select **File > Quit** from the Arduino IDE menus if it is running.
1. Delete the file at the following path:
   - **Windows:**
     ```text
     C:\Users\<username>\.arduinoIDE\settings.json
     ```
     (where `<username>` is your Windows username)
   - **Linux:**
     ```text
     ~/.arduinoIDE/settings.json
     ```
   - **macOS:**
     ```text
     ~/.arduinoIDE/settings.json
     ```
1. Start Arduino IDE.
   🐛 There is an icon on the left side of the editor tabs:
   ![image](https://user-images.githubusercontent.com/8572152/235383157-b21e84c6-9606-4b9c-a7e8-b5c3d8576d0a.png)
1. Open the bottom panel.
   🐛 There is an icon on the left side of the bottom panel tabs:
   ![image](https://user-images.githubusercontent.com/8572152/235383166-feede698-7af1-41e7-9a0f-7c9dfe810c02.png)
1. Run the "**Preferences: Open Settings (UI)**" command.
1. Search for the `workbench.iconTheme` setting.

🐛 `workbench.iconTheme` is set to "**File Icons (Theia)**".

### Expected behavior

The default value of the `workbench.iconTheme` setting is `none`.

### Arduino IDE version

fc4882c (tester build for d66da250a9493b49a42193266269ba6a58ccd8ba)

### Operating system

Windows 11

kittaakos commented:
> ### Expected behavior
> The default value of the `workbench.iconTheme` setting is `none`.

Thanks for noticing it. An upstream change (https://github.com/eclipse-theia/theia/commit/b33ab2cd5e9044c94533822f12d7f9d01a30e822) broke the default behavior. I've adjusted it here: https://github.com/arduino/arduino-ide/pull/2027/commits/7ff070c88baa27545cc7f80c8821a576440a917b.

dankeboy36 commented on arduino-ide-extension/src/electron-browser/electron-app-service.ts:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  registerStartupTasksHandler(
+    handler: (args: unknown) => MaybePromise<void>
```
can be `handler: (tasks: StartupTasks)` from electron api.
should be => void

kittaakos replied:
https://github.com/arduino/arduino-ide/pull/2027/commits/241baca40be0ed10b4d872c65f96d9034190f358

dankeboy36 commented on arduino-ide-extension/src/browser/contributions/startup-tasks-executor.ts:
```diff
@@ -0,0 +1,62 @@
+    this.toDispose.dispose();
+  }
+
+  private async handleStartupTasks(args: unknown): Promise<void> {
```
`args` can be a type of `StartupTasks`

kittaakos replied:
https://github.com/arduino/arduino-ide/pull/2027/commits/241baca40be0ed10b4d872c65f96d9034190f358

kittaakos commented on arduino-ide-extension/src/browser/theia/core/theming.ts:
```diff
@@ -91,7 +102,7 @@ interface ThemeProvider {
  * Returns with a list of built-in themes officially supported by IDE2 (https://github.com/arduino/arduino-ide/issues/1283).
  * The themes in the array follow the following order:
- *  - built-in themes first (in `Light`, `Dark`, `High Contrast`), // TODO -> High Contrast will be split up to HC Dark and HC Light after the Theia version uplift
+ *  - built-in themes first (in `Light`, `Dark`, `High Contrast (Light)`, and `High Contrast (Dark)`), // TODO -> High Contrast will be split up to HC Dark and HC Light after the Theia version uplift
```
https://github.com/arduino/arduino-ide/pull/2027/commits/241baca40be0ed10b4d872c65f96d9034190f358

kittaakos commented on arduino-ide-extension/src/browser/app-service.ts:
```diff
@@ -0,0 +1,13 @@
+  quit(): void;
+  version(): Promise<string>;
+  registerStartupTasksHandler(
+    handler: (args: unknown) => MaybePromise<void>
```
https://github.com/arduino/arduino-ide/pull/2027/commits/241baca40be0ed10b4d872c65f96d9034190f358

kittaakos commented on arduino-ide-extension/src/browser/contributions/startup-tasks-executor.ts:
```diff
@@ -0,0 +1,62 @@
+
+  @postConstruct()
+  protected override init(): void {
+    this.toDispose.push(
```
https://github.com/arduino/arduino-ide/pull/2027/commits/241baca40be0ed10b4d872c65f96d9034190f358

AlbyIanna commented:
### The style of the release notes in the  IDE updater dialog is broken
I've  edited the version of the app in the package.json so that the updater dialog would show up, and this is the result:
<img width="610" alt="Screenshot 2023-05-03 at 16 42 32" src="https://user-images.githubusercontent.com/6939054/235953106-a0b00ae6-8b8e-49ce-adeb-846b42ca1833.png">

This is how it should appear:
<img width="608" alt="Screenshot 2023-05-03 at 16 43 07" src="https://user-images.githubusercontent.com/6939054/235953212-4d5a6a12-cae5-44a8-9bfc-3ea085c3c7eb.png">

kittaakos commented:
> The style of the release notes in the IDE updater dialog is broken

Thanks for checking it.

The breaking change was https://github.com/eclipse-theia/theia/commit/1b5ff9ee459df14cedc0e8266dd02dae93fcd1bf#diff-d8d45a890507a01141c010ad4a6891edf2ae727cfa6dfe604cebbd667812cccbR68. I fixed it 
45aeb7f.

per1234 reviewed (changes requested):
**UPDATE: Fixed by https://github.com/arduino/arduino-ide/pull/2027/commits/6adde6b631350af4e83b3a23cc3a43f36b7bb33a**

<s>
## **Edit > Paste** menu item is non-functional

### Describe the problem

🐛 Selecting the **Edit > Paste** menu item does not have any effect.

### To reproduce

1. Select **File > New Sketch** from the Arduino IDE menus.
1. Select the text `// put your setup`
1. Select **Edit > Copy** from the Arduino IDE menus.
1. Move the cursor to another location in the sketch.
1. Select **Edit > Paste** from the Arduino IDE menus.

🐛 The copied text is not pasted.

The IDE displays a warning notification:

> **⚠ Please use the browser's paste command or shortcut.**

### Expected behavior

Select **Edit > Paste** from the Arduino IDE menus pastes the content of the clipboard into the editor at the cursor location.

### Arduino IDE version

0f1bf56 (tester build for https://github.com/arduino/arduino-ide/pull/2027/commits/45aeb7f198a0d8907b1b80157fc2d0f12b6d00cc)

### Operating system

- Windows 11
- Ubuntu 22.04

### Additional context

The fault does not occur when using the build from the `main` branch (51f69f6a5930ed1c947ac708b2d7280ea3e0cd57)

I am not able to reproduce the fault on my macOS machine.

---

The fault also occurs when using the "**Paste**" item of the editor context menu.

Pasting via the <kbd>Ctrl</kbd>+<kbd>V</kbd> keyboard shortcut works as expected, so the fault is specific to the **Edit > Paste** menu item.

--- 

I notice no keyboard shortcut is listed for **Paste** in the **Edit** menu:

![image](https://user-images.githubusercontent.com/8572152/236375328-141b5256-d528-4aa6-b6f2-f8cd1cece9ff.png)

compare to 51f69f6a5930ed1c947ac708b2d7280ea3e0cd57:

![image](https://user-images.githubusercontent.com/8572152/236375353-15131207-106b-4d28-b7aa-c58feed7db22.png)

The shortcut is also not defined for the "**Paste**" command in the "**Keyboard Shortcuts**" view:

![image](https://user-images.githubusercontent.com/8572152/236375378-5b9e106c-45fe-436e-9970-9f05e2592ca0.png)

compare to 51f69f6a5930ed1c947ac708b2d7280ea3e0cd57:

![image](https://user-images.githubusercontent.com/8572152/236375397-8944e9b2-4fba-4adc-93bb-4afc341c5a42.png)

However, the fault still occurs even after I assign the <kbd>Ctrl</kbd>+<kbd>V</kbd> keybinding to the "**Paste**" command.
</s>

per1234 reviewed (changes requested):
**UPDATE: fixed by https://github.com/arduino/arduino-ide/pull/2027/commits/6adde6b631350af4e83b3a23cc3a43f36b7bb33a**

<s>
## **Show Sketch Folder** has no effect

### Describe the problem

The sketch folder is not opened when you select **Sketch > Show Sketch Folder** from the Arduino IDE menus.

### To reproduce

1. Select **Sketch > Show Sketch Folder** from the Arduino IDE menus

🐛 The folder of the sketch is not opened in the file browser.

### Expected behavior

The folder of the sketch currently open in the IDE window is opened in the file browser when **Sketch > Show Sketch Folder** is selected from the Arduino IDE menus.

### Arduino IDE version

0f1bf56 (tester build for 45aeb7f)

### Operating system

Windows 11

### Additional context

The fault does not occur when using the build from the `main` branch (51f69f6a5930ed1c947ac708b2d7280ea3e0cd57)

The fault does not occur on my Linux and macOS machines.

---

Nothing is printed to the logs when I do a **Show Sketch Folder** operation.
</s>

per1234 reviewed (commented):
## Non-functional language server-related items added to editor context menu

This is an expansion of the existing problem tracked at https://github.com/arduino/arduino-ide/issues/1394

Maybe the conclusion will be the same:

https://github.com/arduino/arduino-ide/issues/1394#issuecomment-1236648266

> The LS should use them. Instead of removing the context menu items

but I thought it was at least worth mentioning because we now have a significant number of non-functional items and I am not aware of any work, plans, or investigation related to bringing them into a functional state.

### Describe the problem

New items were added to the editor context menu:

- Go to Declaration
- Peek > Peek Declaration
- Go to Type Definition
- Peek > Peek Type Definition
- Go to Implementations
- Peek > Peek Implementations
- Go to References
- Peek > Peek References

🐛 None of the newly added items are functional.

### To reproduce

1. Create a sketch with the following content:
   ```cpp
   typedef int foo_t;
   foo_t bar();
   foo_t bar() {
     return 42;
   }
   void setup() {
     bar();
   }
   ```
1. Select **Tools > Board > Arduino AVR Boards > Arduino Uno** from the Arduino IDE menus.
1. Wait for the the "indexing" process to finish, as indicated at the left side of the status bar.
1. Right click on the `bar` on line 7.
1. Select "**Go to Declaration**" from the context menu.
   🐛 Nothing happens (expected result is that the cursor would move to the `bar` function declaration at line 2)
1. Right click on the `bar` on line 7.
1. Select "**Peek > Peek Declaration**" from the context menu.
   🐛 Nothing happens (expected result is that a peek of the `bar` declaration would open)
1. Right click on the `foo_t` on line 2.
1. Select "**Go to Type Definition**" from the context menu.
   🐛 Nothing happens (expected result is that the cursor would move to the `foo_t` type definition at line 1)
1. Right click on the `foo_t` on line 2.
1. Select "**Peek > Peek Type Definition**" from the context menu.
   🐛 Nothing happens (expected result is that a peek of the `foo_t` type definition would open)
1. Right click on the `bar` on line 2.
1. Select "**Go to References**" from the context menu.
   🐛 Nothing happens (expected result is that a peek would open with all references to `bar`)
1. Right click on the `bar` on line 2.
1. Select "**Peek > Peek References**" from the context menu.
   🐛 Nothing happens (expected result is that a peek would open with all references to `bar`)

### Expected behavior

All editor context menu items are functional.

### Arduino IDE version

0f1bf56 (tester build for 45aeb7f)

### Operating system

Windows 11

### Additional context

The non-functional menu items are not present when using the build from the `main` branch (51f69f6a5930ed1c947ac708b2d7280ea3e0cd57)

---

I did not test "**Go to Implementations**" and "**Peek Implementations**" because I don't know what "implementation" means in the context of a C++ program.

kittaakos commented:
We must update the minimum Node.js version. It's `>=16.13.0`:


```
yarn install v1.22.5
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
warning @puppeteer/browsers@0.5.0: Invalid bin entry for "@puppeteer/browsers" (in "@puppeteer/browsers").
info fsevents@2.3.2: The platform "win32" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
error @npmcli/arborist@6.2.3: The engine "node" is incompatible with this module. Expected version "^14.17.0 || ^16.13.0 || >=18.0.0". Got "16.0.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

kittaakos commented:
Currently, I cannot even build IDE2 from the source on Windows. Theia does not work either: https://github.com/eclipse-theia/theia/issues/12486

kittaakos commented:
> **Edit > Paste** menu item is non-functional

Upstream: https://github.com/eclipse-theia/theia/issues/12487

I will investigate. Thank you for noticing it 🙏

AlbyIanna commented on arduino-ide-extension/src/browser/dialogs/ide-updater/ide-updater-component.tsx:
```diff
@@ -1,10 +1,15 @@
 import ProgressBar from '../../components/ProgressBar';
 
+const ReactMarkdown = React.lazy<React.ComponentType<Options>>(
+  // @ts-expect-error see above
```
I don't think this line is needed, as I'm getting this compilation error:

```

src/browser/dialogs/ide-updater/ide-updater-component.tsx:9:3 - error TS2578: Unused '@ts-expect-error' directive.

9   // @ts-expect-error see above
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

[12:50:39 PM] Found 1 error. Watching for file changes.
```

I've tried to remove it and the markdown in the dialog seems to be displayed correctly.

kittaakos replied:
> as I'm getting this compilation error:

in Code or with `tsc`? Without it, it does not compile for me in Code.

AlbyIanna replied:
I just pulled the latest changes and ran `yarn` again and it worked 👍

kittaakos commented:
> We must update the minimum Node.js version. It's `>=16.13.0`:

Done. It's updated to 16.14.0.

> **Show Sketch Folder** has no effect

I have fixed it and verified it from the latest build. Please re-check

> **Edit > Paste** menu item is non-functional

I added a workaround. It worked for me with the latest build on Windows.


My build:

Version: 2.1.1-snapshot-e2539a8
Date: 2023-05-05T10:08:57.749Z
CLI Version: 0.32.2

Copyright © 2023 Arduino SA

AlbyIanna commented:
## Some strings are not getting localised

A couple of examples:

### 2.1.0
<details>
<summary>
<b>Edit</b> menu (click to see)
</summary>
<img width="1011" alt="Screenshot 2023-05-05 at 16 57 40" src="https://user-images.githubusercontent.com/6939054/236494315-c168b710-d2b9-45a2-b1ad-afdf9078ad01.png">

</details>

<details>
<summary>
Preferences (click to see)
</summary>
<img width="489" alt="Screenshot 2023-05-05 at 16 49 59" src="https://user-images.githubusercontent.com/6939054/236494764-c0417077-8c73-42d5-ac63-61ba9d146467.png">

</details>





### 2.1.1-snapshot-53e853a
<details>
<summary>
<b>Edit</b> menu (click to see)
</summary>

</details>

<details>
<summary>
Preferences (click to see)
</summary>

`Modified by` should be `Modifié dans`.


<img width="489" alt="Screenshot 2023-05-05 at 16 50 43" src="https://user-images.githubusercontent.com/6939054/236494759-adb51a70-0bfe-44df-afe0-297922bd1f1b.png">

</details>

kittaakos commented:
> ## Some strings are not getting localised
> A couple of examples:

Very nice 👍 

I have updated the VS Code extension language pack versions to match Theia's supported VS Code API version: https://github.com/eclipse-theia/theia/blob/v1.37.0/dev-packages/application-package/src/api.ts#L21. See the changes:

<img width="557" alt="Screen Shot 2023-05-06 at 17 07 38" src="https://user-images.githubusercontent.com/1405703/236632562-76e16303-e22c-45d3-abfa-97376a52b635.png">

<img width="483" alt="Screen Shot 2023-05-06 at 17 04 31" src="https://user-images.githubusercontent.com/1405703/236632574-7b509b6b-5271-4374-b010-0652298bf096.png">


I also provided a better workaround for the cut/copy/paste. I will wait for a Windows build and do a double check, but it works on macOS as it was with 2.1.0

kittaakos commented:
> I have updated the VS Code extension language pack versions to match Theia's supported VS Code API version:

Some supported languages do not have a more recent extension version available. We must check why and what are the implications of using older language packs with Theia using VS Code 1.72.x APIx:

`bg`, `hu`, `nl`, and `uk` extensions do not have a more recent version available.

per1234 reviewed (dismissed):
**UPDATE: resolved by https://github.com/arduino/arduino-ide/pull/2027/commits/bcb7f7df8a2384438f2969c217b5c4aa0ccbc8c4**

<s>
## Some UI actions have no effect if performed soon after opening window

After opening an IDE window, some sort of an initialization process continues even after the UI appears in the window.

🐛 Some actions have no effect if the user performs them during this time window.

### To reproduce

1. Select **File > New Sketch** from the Arduino IDE menus.
1. Wait until the UI populates in the new window.
1. Immediately* use the mouse to select **File > New Sketch** from the Arduino IDE menus.
   On Windows, this must be done while the mouse pointer has the "working in background" style while hovered over the IDE window:
   ![image](https://user-images.githubusercontent.com/8572152/236657783-b222b2e5-4b05-418a-97c3-e75040b18b64.png)

🐛 A new window does not open.

### Expected behavior

UI elements are usable whenever they are exposed to the user.

### Arduino IDE version

0fb9a06 (tester build for https://github.com/arduino/arduino-ide/pull/2027/commits/ed600489da57e655908bee234e4f34585920dd54)

### Operating system

- Windows
- macOS

### Operating system version

- Windows 11
- macOS Ventura

### Additional context

The fault does not occur when using the build from the `main` branch (964ea3b)

---

I used **New Sketch** in the demo, but the fault occurs with other UI operations as well.

---

I was not able to reproduce the fault on my Linux machine.

The fault does not occur when a keyboard shortcut (e.g., <kbd>Ctrl</kbd>+<kbd>N</kbd>) or even keyboard navigation of the GUI is used to trigger the operation instead of the mouse.

The fault does not occur if I make a keystroke before the UI action.

---

\* On my high spec PC the window is ~5 seconds. It will likely be significantly longer for users with lower performance systems (and thus those users will be statistically more likely to be impacted by the regression).
</s>

kittaakos commented:
> I have updated the VS Code extension language pack versions to match Theia's supported VS Code API version:

Something needs to be corrected. We might need to pick another version like we did last time. Related: https://github.com/arduino/arduino-ide/pull/1449

<img width="334" alt="Screen Shot 2023-05-08 at 09 05 15" src="https://user-images.githubusercontent.com/1405703/236757679-ca002075-d07d-499f-beed-f20d1dc762c7.png">

kittaakos commented:
> Non-functional language server-related items added to editor context menu

Fixed. Also, #1394 should be fixed.


<img width="644" alt="Screen Shot 2023-05-08 at 20 10 51" src="https://user-images.githubusercontent.com/1405703/236899376-b85e8857-8cc9-4ce4-93f4-dbda708cf918.png">

Please mention in the review if something else needs to be removed. Thanks!

----

> > I have updated the VS Code extension language pack versions to match Theia's supported VS Code API version:
> 
> Something needs to be corrected. We might need to pick another version like we did last time. Related: #1449
> 
> <img alt="Screen Shot 2023-05-08 at 09 05 15" width="334" src="https://user-images.githubusercontent.com/1405703/236757679-ca002075-d07d-499f-beed-f20d1dc762c7.png">

Switched to 1.17.0 language extension.

<img width="568" alt="Screen Shot 2023-05-08 at 17 08 08" src="https://user-images.githubusercontent.com/1405703/236899681-e67eefa5-cd77-47e0-8428-49c462ff5794.png">
<img width="490" alt="Screen Shot 2023-05-08 at 17 07 53" src="https://user-images.githubusercontent.com/1405703/236899687-b1a7d5e4-e7b4-405a-832c-3d3bd05d7bc0.png">


----

> Some UI actions have no effect if performed soon after opening window

The ugliest workaround is available for verification, but it works. Please verify it.


https://user-images.githubusercontent.com/1405703/236900279-e83bf687-8b85-4349-a8a2-b62fff49399a.mp4

----

Please revisit the behavior with the most recent changes. Thanks a lot for your persistence. This PR takes way longer than I anticipated.

kittaakos commented:
> Switched to 1.17.0 language extension.

This still applies 👇 . From https://github.com/arduino/arduino-ide/pull/2027#issuecomment-1537162699:

> Some supported languages do not have a more recent extension version available. We must check why and what are the implications of using older language packs with Theia using VS Code 1.72.x APIx:
> 
> `bg`, `hu`, `nl`, and `uk` extensions do not have a more recent version available.

kittaakos commented on arduino-ide-extension/src/browser/dialogs/ide-updater/ide-updater-dialog.tsx:
```diff
@@ -76,11 +74,15 @@ export class IDEUpdaterDialog extends ReactDialog<UpdateInfo | undefined> {
   }
 
+  private readonly openExternal = (url: string) =>
+    this.windowService.openNewWindow(url, { external: true });
```
> @kittaakos what about adding validation on the URL parameter to accept only `https` and `http` URLs?
> It helps to follow a security-by-design strategy in case of code evolution

I appreciate your input. This is a great idea but unrelated to this PR and the Theia version update. I plan to keep this the same in the context of this PR. Thanks for your understanding!

----

I remember bringing up this topic in one of the internal chats, please open a follow-up either in [arduino/arduino-ide](https://github.com/arduino/arduino-ide/) or directly to [eclipse-theia/theia/](https://github.com/eclipse-theia/theia).

There are a few related issues in Theia:
 - [vscode: Support trusted commands in markdown and webviews](https://github.com/eclipse-theia/theia/issues/12079)
 - [[vscode] Implement full Workspace Trust behavior](https://github.com/eclipse-theia/theia/issues/12318)

Here is the corresponding VS Code documentation: https://code.visualstudio.com/docs/editor/workspace-trust
Here is how VS Code handles trusted domains: https://github.com/microsoft/vscode/blob/a54b497150197f773f485038f44af02772c5a016/src/vs/workbench/contrib/url/browser/trustedDomains.ts

kittaakos replied:
> adding validation on the URL

For completeness, this is the expected place for having such URL validation in IDE2:

https://github.com/arduino/arduino-ide/blob/964ea3bc0cb51c87592757ca7b8b73002c475033/arduino-ide-extension/src/electron-main/theia/electron-main-window-service.ts#L24-L26

kittaakos replied:
This topic is not tracked at https://github.com/arduino/arduino-ide/issues/2056.

I am resolving this thread.

AlbyIanna reviewed (approved):
All my comments were addressed! I couldn't find any other issue and the code looks good to me. Thank, Akos!

kittaakos commented:
> Some supported languages do not have a more recent extension version available. We must check why and what are the implications of using older language packs with Theia using VS Code 1.72.x APIx:
> 
> `bg`, `hu`, `nl`, and `uk` extensions do not have a more recent version available.


>  We must check why

 - there is no more recent version of the language packs for `bg` and `hu` in the VS Code Marketplace, so Open SVX cannot make a copy.
    - https://marketplace.visualstudio.com/items?itemName=MS-CEINTL.vscode-language-pack-bg
    - https://marketplace.visualstudio.com/items?itemName=MS-CEINTL.vscode-language-pack-hu
 - `nl` and `uk` no longer exist in the VS Code Marketplace.

kittaakos commented:
Great reviews! Thanks a lot for your time and patience. 🙌
